jax/api.py

```diff
@@ -2318,8 +2318,14 @@ def __init__(self, shape, dtype):
   >>> print(out.dtype)
   float32
   """
+  def dtype(x):
+    try:
+      return dtypes.result_type(x)
+    except ValueError:
+      return dtypes.result_type(getattr(x, 'dtype'))
```

Comment:
I'm not sure this is the place for this kind of logic. JAX already handles some of these cases by delegating to `np.result_type`, which checks for the `dtype` attribute:
```python
import numpy as np
import jax.numpy as jnp

class MyArray:
  def __init__(self, dtype=np.int32):
    self.dtype = dtype

print(jnp.result_type(MyArray()))
# int32
```
This doesn't work for `EasyDict` because `numpy.result_type` fails to recognize it.

I think we should try to understand why numpy fails this test before working around it in the jax wrapper. And if the workaround is warranted, IMO it would be better to do it at the location that we fall-back to `np.result_type`, which I believe is here: https://github.com/google/jax/blob/c2455d97e95f5c3c31a96571a1236b1f786ad93d/jax/dtypes.py#L319-L322

Reply:
...to fully support this we'd probably also have to add similar logic to `jnp.dtype`, which currently is just an alias to `np.dtype`, which fails in the same way.

Reply:
Actually, I already tried something like that: the earlier commit d99cba put this logic in dtypes.py, but that led to failures in existing tests for subtle reasons.

The reason the logic makes sense here is that `jax.eval_shape` explicitly says in its docstring that it supports its inputs being duck-typed this way:

> Since only the ``shape`` and ``dtype`` attributes are accessed, only values that duck-type arrays are required, rather than real ndarrays.

Other parts of our API, like `jnp.dtype`, don't make that promise, so we don't actually need to generalize the functions in dtypes.py. (If anything, it'd be surprising to make `jnp` API functions differ from their `np` cousins in this way!)

Reply:
In other words, the goal of this PR is not to change `jnp.result_type` or the logic in dtypes.py at all. We just want to fix `jax.eval_shape`.

Reply:
OK, that makes sense!